File: generator-ui/src/types/generator.ts
/**
 * Generator UI Types
 */

// 시나리오 카테고리
export type ScenarioCategory =
  | 'mes_production'
  | 'mes_equipment'
  | 'mes_quality'
  | 'mes_material'
  | 'mes_system'
  | 'erp_sales'
  | 'erp_purchase'
  | 'erp_inventory'
  | 'erp_accounting'
  | 'erp_hr'
  | 'erp_master';

// 시나리오 정보
export interface Scenario {
  id: string;
  name: string;
  description: string;
  category: ScenarioCategory;
  enabled: boolean;
  ai_use_cases: string[];
  trigger_type: 'scheduled' | 'condition' | 'random' | 'always';
  start_date?: string;
  duration_days?: number;
}

// 생성 설정
export interface GeneratorConfig {
  start_date: string;
  end_date: string;
  tenant_id: string;
  random_seed: number;
  enabled_scenarios: string[];
  output_format: 'json' | 'database';
}

// 생성 작업 상태
export type JobStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';

// 생성 작업
export interface GeneratorJob {
  id: string;
  status: JobStatus;
  config: GeneratorConfig;
  progress: GeneratorProgress;
  created_at: string;
  started_at?: string;
  completed_at?: string;
  error_message?: string;
}

// 진행 상태
export interface GeneratorProgress {
  current_day: number;
  total_days: number;
  current_date: string;
  percentage: number;
  current_module?: string;
  records_generated: RecordCounts;
}

// 레코드 카운트
export interface RecordCounts {
  mes: {
    production_orders: number;
    production_results: number;
    equipment_status: number;
    quality_inspections: number;
    defect_records: number;
    material_consumption: number;
  };
  erp: {
    sales_orders: number;
    purchase_orders: number;
    inventory_transactions: number;
    journal_entries: number;
    attendance_records: number;
  };
}

// 생성 요약
export interface GeneratorSummary {
  mes: {
    production: ModuleSummary;
    equipment: ModuleSummary;
    quality: ModuleSummary;
    material: ModuleSummary;
  };
  erp: {
    sales: ModuleSummary;
    purchase: ModuleSummary;
    inventory: ModuleSummary;
    accounting: ModuleSummary;
    hr: ModuleSummary;
  };
}

export interface ModuleSummary {
  [key: string]: number | string;
}

// WebSocket 메시지
export interface WSMessage {
  type: 'progress' | 'completed' | 'error' | 'log';
  job_id: string;
  data: GeneratorProgress | GeneratorSummary | string;
  timestamp: string;
}

// API 응답
export interface ApiResponse<T> {
  success: boolean;
  data?: T;
  error?: string;
}

// 시나리오 카테고리 메타데이터
export const SCENARIO_CATEGORIES: Record<ScenarioCategory, { label: string; icon: string; color: string }> = {
  mes_production: { label: '생산관리', icon: 'Factory', color: 'blue' },
  mes_equipment: { label: '설비관리', icon: 'Cog', color: 'orange' },
  mes_quality: { label: '품질관리', icon: 'CheckCircle', color: 'green' },
  mes_material: { label: '자재관리', icon: 'Package', color: 'purple' },
  mes_system: { label: '시스템/환경', icon: 'Settings', color: 'gray' },
  erp_sales: { label: '영업관리', icon: 'ShoppingCart', color: 'indigo' },
  erp_purchase: { label: '구매관리', icon: 'Truck', color: 'yellow' },
  erp_inventory: { label: '재고관리', icon: 'Warehouse', color: 'teal' },
  erp_accounting: { label: '회계관리', icon: 'Calculator', color: 'red' },
  erp_hr: { label: '인사급여', icon: 'Users', color: 'pink' },
  erp_master: { label: '기준정보', icon: 'Database', color: 'cyan' },
};

File: api/schemas/generator.py
"""
Generator API Schemas
Pydantic models for data generator endpoints
"""
from datetime import date, datetime
from typing import Dict, List, Optional, Any
from enum import Enum
from pydantic import BaseModel, Field


class ScenarioCategory(str, Enum):
    MES_PRODUCTION = "mes_production"
    MES_EQUIPMENT = "mes_equipment"
    MES_QUALITY = "mes_quality"
    MES_MATERIAL = "mes_material"
    MES_SYSTEM = "mes_system"
    ERP_SALES = "erp_sales"
    ERP_PURCHASE = "erp_purchase"
    ERP_INVENTORY = "erp_inventory"
    ERP_ACCOUNTING = "erp_accounting"
    ERP_HR = "erp_hr"
    ERP_MASTER = "erp_master"


class TriggerType(str, Enum):
    SCHEDULED = "scheduled"
    CONDITION = "condition"
    RANDOM = "random"
    ALWAYS = "always"


class JobStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class OutputFormat(str, Enum):
    JSON = "json"
    DATABASE = "database"


# Scenario schemas
class ScenarioBase(BaseModel):
    id: str
    name: str
    description: str
    category: ScenarioCategory
    enabled: bool = True
    ai_use_cases: List[str] = []
    trigger_type: TriggerType
    start_date: Optional[str] = None
    duration_days: Optional[int] = None


class ScenarioResponse(ScenarioBase):
    pass


class ScenarioUpdate(BaseModel):
    enabled: bool


# Generator config schemas
class GeneratorConfigBase(BaseModel):
    start_date: str = Field(default="2024-07-01", description="시뮬레이션 시작일")
    end_date: str = Field(default="2024-12-31", description="시뮬레이션 종료일")
    tenant_id: str = Field(default="T001", description="테넌트 ID")
    random_seed: int = Field(default=42, description="Random seed")
    enabled_scenarios: List[str] = Field(default_factory=list, description="활성화된 시나리오 ID 목록")
    output_format: OutputFormat = Field(default=OutputFormat.JSON, description="출력 형식")


class GeneratorConfigResponse(GeneratorConfigBase):
    pass


class GeneratorConfigUpdate(BaseModel):
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    tenant_id: Optional[str] = None
    random_seed: Optional[int] = None
    enabled_scenarios: Optional[List[str]] = None
    output_format: Optional[OutputFormat] = None


# Record counts
class MESRecordCounts(BaseModel):
    production_orders: int = 0
    production_results: int = 0
    equipment_status: int = 0
    quality_inspections: int = 0
    defect_records: int = 0
    material_consumption: int = 0


class ERPRecordCounts(BaseModel):
    sales_orders: int = 0
    purchase_orders: int = 0
    inventory_transactions: int = 0
    journal_entries: int = 0
    attendance_records: int = 0


class RecordCounts(BaseModel):
    mes: MESRecordCounts = Field(default_factory=MESRecordCounts)
    erp: ERPRecordCounts = Field(default_factory=ERPRecordCounts)


# Progress schemas
class GeneratorProgress(BaseModel):
    current_day: int = 0
    total_days: int = 0
    current_date: str = ""
    percentage: float = 0.0
    current_module: Optional[str] = None
    records_generated: RecordCounts = Field(default_factory=RecordCounts)


# Job schemas
class GeneratorJobCreate(GeneratorConfigBase):
    pass


class GeneratorJobResponse(BaseModel):
    id: str
    status: JobStatus
    config: GeneratorConfigBase
    progress: GeneratorProgress
    created_at: datetime
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    error_message: Optional[str] = None

    class Config:
        from_attributes = True


# Summary schemas
class ModuleSummary(BaseModel):
    total_records: int = 0
    tables: Dict[str, int] = Field(default_factory=dict)


class GeneratorSummary(BaseModel):
    mes: Dict[str, ModuleSummary] = Field(default_factory=dict)
    erp: Dict[str, ModuleSummary] = Field(default_factory=dict)
    total_records: int = 0
    duration_seconds: float = 0.0


# WebSocket message
class WSMessageType(str, Enum):
    PROGRESS = "progress"
    COMPLETED = "completed"
    ERROR = "error"
    LOG = "log"


class WSMessage(BaseModel):
    type: WSMessageType
    job_id: str
    data: Any
    timestamp: datetime = Field(default_factory=datetime.now)


# API Response wrapper
class ApiResponse(BaseModel):
    success: bool = True
    data: Optional[Any] = None
    error: Optional[str] = None
